### ISSUE TITLE
RemoteWebDriver cannot connect to Selenium Grid 4

### ISSUE THREAD

**Issue by stawr93** [SeleniumHQ/selenium#9277]
## 🐛 Bug Report

The exception is thrown on the RemoteWebDriver object instantiation when application targeting .Net Framework 4.8.

## To Reproduce

Run following code in an application targeting net48:

```
var chromeOpts = new ChromeOptions(); 
var driver = new RemoteWebDriver(new Uri("http://selenium-grid:4444/wd/hub/"), chromeOpts); 
driver.Navigate().GoToUrl("http://application-url"); 
driver.Quit();
```

You'll get exception like:

```
Unhandled Exception: OpenQA.Selenium.WebDriverException: The HTTP request to the remote WebDriver server for URL http://selenium-grid:4444/wd/hub/session timed out after 60 seco
nds. ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at OpenQA.Selenium.Remote.HttpCommandExecutor.<MakeHttpRequest>d__30.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at OpenQA.Selenium.Remote.HttpCommandExecutor.Execute(Command commandToExecute)
   --- End of inner exception stack trace ---
   at OpenQA.Selenium.Remote.HttpCommandExecutor.Execute(Command commandToExecute)
   at OpenQA.Selenium.Remote.RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
   at OpenQA.Selenium.Remote.RemoteWebDriver.StartSession(ICapabilities desiredCapabilities)
   at OpenQA.Selenium.Remote.RemoteWebDriver..ctor(ICommandExecutor commandExecutor, ICapabilities desiredCapabilities)
   at OpenQA.Selenium.Remote.RemoteWebDriver..ctor(Uri remoteAddress, ICapabilities desiredCapabilities, TimeSpan commandTimeout)
   at OpenQA.Selenium.Remote.RemoteWebDriver..ctor(Uri remoteAddress, ICapabilities desiredCapabilities)
   at OpenQA.Selenium.Remote.RemoteWebDriver..ctor(Uri remoteAddress, DriverOptions options)
   at ConsoleApplication1.Program.Main(String[] args) in C:\Projects\Test\SeleniumTest\ConsoleApplication1\ConsoleApplication1\Program.cs:line 13
```

## Expected behavior

A RemoteWebDriver created successfully.

## Test script or set of commands reproducing this issue

See above.

## Environment

OS: Windows 10 
Browser: any
Browser version: any
Browser Driver version: any
Language Bindings version: .NET Selenium.WebDriver 4.0.0-beta1
Selenium Grid version (if applicable): docker image `selenium/hub:4.0.0-beta-2-prerelease-20210312`

**Comment by diemol** [comment]
Could you please share the logs of the Docker containers? Also, how are you starting the containers?

**Comment by stawr93** [comment]
I'm using docker compose from selenium-docker: https://github.com/SeleniumHQ/docker-selenium/blob/trunk/docker-compose-v3.yml. Ports are exposed correctly - I'm able to create a session in the same app using selenium.webdriver v3.12.

selenium hub logs have no any session creation records. it seems like the request does not reach the hub. I was able to successfully create a session in the application that are targeting .net core 2.1.

My guess is an issue might be related to HttpClient.

**Comment by omkarnathsingh** [comment]
I am getting below error when trying to create session when I use Selenium 4 beta 2 in standalone mode, I have updated the maven dependency to use selenium-java 4, beta 2 as well. It works when using server version 3.141.59, but not 4. Can someone suggest a solution.

```
[main] WARN  SeleniumBase : Couldn't get Driver [[ http://localhost:4444/wd/hub]]org.openqa.selenium.SessionNotCreatedException: Could not start a new session. Response code 500. Message: Could not start a new session. No Node supports the required capabilities: Capabilities {browserName: chrome, goog:chromeOptions: {args: [test-type, -disable-popup-blocking, --disable-notifications, chrome.switches, --disable-extensions, start-maximized, --disable-dev-shm-usage, --lang=en-GB, --ignore-certificate-errors], excludeSwitches: [enable-automation], extensions: [], prefs: {credentials_enable_service: false, download.default_directory: C:\Development\Download, download.directory_upgrade: true, download.prompt_for_download: false, profile.content_settings.exceptions.clipboard: {[*.],*: {last_modified: 1616397209285, setting: 1}}}}, pageLoadStrategy: normal, platformName: windows}
Build info: version: '4.0.0-beta-2', revision: 'Unknown'
System info: os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '11.0.5'
Driver info: driver.version: unknown
Build info: version: '4.0.0-beta-2', revision: '4ec0cfc112'
System info: host: os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '11.0.5'
```

**Comment by diemol** [comment]
@omkarnathsingh your comment is off-topic as this issue is about the C# bindings. If you are having an issue please open a new one.

**Comment by omkarnathsingh** [comment]
Sorry for offtopic, will make a new one with details.

**Comment by pujagani** [comment]
Thank you for sharing the details and the follow-up. 
I have tried the same example with this docker file -> https://github.com/SeleniumHQ/docker-selenium/blob/d14e2dabe125f4b41a87f5cee34ebed96717b112/docker-compose-v3.yml 

I see the following logs, as expected : 

```
selenium-hub    | 07:51:37.855 INFO [Distributor.newSession] - Session request received by the distributor: 
selenium-hub    |  {
selenium-hub    |   "desiredCapabilities": {
selenium-hub    |     "browserName": "chrome",
selenium-hub    |     "goog:chromeOptions": {
selenium-hub    |     }
selenium-hub    |   },
selenium-hub    |   "capabilities": {
selenium-hub    |     "firstMatch": [
selenium-hub    |       {
selenium-hub    |         "browserName": "chrome",
selenium-hub    |         "goog:chromeOptions": {
selenium-hub    |         }
selenium-hub    |       }
selenium-hub    |     ]
selenium-hub    |   }
selenium-hub    | }
chrome_1        | Starting ChromeDriver 89.0.4389.23 (61b08ee2c50024bab004e48d2b1b083cdbdac579-refs/branch-heads/4389@{#294}) on port 14483
chrome_1        | Only local connections are allowed.
chrome_1        | Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
chrome_1        | ChromeDriver was started successfully.
chrome_1        | [1617263498.095][SEVERE]: bind() failed: Cannot assign requested address (99)
chrome_1        | 07:51:39.332 INFO [ProtocolHandshake.createSession] - Detected dialect: W3C
selenium-hub    | 07:51:39.431 INFO [Distributor.newSession] - Session created by the distributor. Id: 27a59ef50147e70036aac572bc07088f, Caps: Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 89.0.4389.82, chrome: {chromedriverVersion: 89.0.4389.23 (61b08ee2c5002..., userDataDir: /tmp/.com.google.Chrome.VIxUgA}, goog:chromeOptions: {debuggerAddress: localhost:41269}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: linux, proxy: {}, se:cdp: http://172.22.0.6:5555/sess..., setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:largeBlob: true, webauthn:virtualAuthenticators: true}
selenium-hub    | 07:51:41.267 INFO [LocalSessionMap.lambda$new$0] - Deleted session from local session map, Id: 27a59ef50147e70036aac572bc07088f
```
The .csproj file for the project is as follows:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net4.8</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Selenium.WebDriver" Version="4.0.0-beta1" />
    <PackageReference Include="Selenium.WebDriver.ChromeDriver" Version="89.0.4389.2300" />
  </ItemGroup>
</Project>

```

After starting the docker containers, please ping "http://localhost:4444/status" to see if the hub is reachable. 

I am unable to reproduce the issue with the same setup. Please provide more details or context to help address the issue.

**Comment by pujagani** [comment]
`OpenQA.Selenium.WebDriverException: The HTTP request to the remote WebDriver server for URL http://selenium-grid:4444/wd/hub/session timed out after 60 seconds.`

I suspect this will happen if the previously the code was executed and it errored out but the `driver.quit()` was not called, hence the session is not deleted. When the request is made again since the previous session exists occupying the Chrome slot, the request is retried and it times out due to the HTTP client defaults. 

On the client-side this can be fixed by using the example below:
 ```
class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Started");
            var chromeOpts = new ChromeOptions();

            var driver = new RemoteWebDriver(new Uri("http://localhost:4444/wd/hub/"), chromeOpts.ToCapabilities(), TimeSpan.FromMinutes(5));

            driver.Navigate().GoToUrl("http://www.google.com");

            driver.Quit();

            Console.WriteLine("Finish");
        }
    }
```

On the server-side, the slot will be reserved for default of 5 minutes. If no activity happens in the browser for 5 mins then it will free up. This the controlled by the node session timeout property. 

To configure the value, refer to the example below: 

```
version: '3.7'

services:
  hub4:
    image: selenium/hub:4.0.0-beta-2-20210317
    ports:
      - "4442:4442"
      - "4443:4443"
      - "4444:4444"

  chromenode:
    image: selenium/node-chrome:4.0.0-beta-2-20210317
    volumes:
      - /dev/shm:/dev/shm
    depends_on:
      - hub4
    environment:
      - SE_EVENT_BUS_HOST=hub4
      - SE_EVENT_BUS_PUBLISH_PORT=4442
      - SE_EVENT_BUS_SUBSCRIBE_PORT=4443
      - SE_NODE_SESSION_TIMEOUT=180
    ports:
      - "6900:5900"
```

The `NODE_SESSION_TIMEOUT`in seconds can be changed as needed so that a hanging session does not lead to timeout for other requests. 

Here, the hanging session will be removed after 3 mins of inactivity and the slot will free up and the session will be deleted. 
Since the client-side timeout is 5 minutes, the request will be successful since the slot is available.

**Comment by BoBtSoL** [comment]
I have exactly the same issue as stawr93 described. It fails in .NET Framework 4.8, but works in .NET core.

**Comment by Dragonjester** [comment]
facing the same issue, as described by stawr93 😢

**Comment by AceOfSpades29** [comment]
Facing same issue, using standalone version of Selenium 4 new project file targeting .NET 4.8. No configuration 

standalone command 
java -jar .\selenium-server-4.0.0-rc-1.jar standalone

create new C# unit test project targeting 4.8

code inside test method with (/wd/hub and without /wd/hub):
```
var chromeOptions = new ChromeOptions();
           
            IWebDriver driver = new RemoteWebDriver(new Uri("http://localhost:4444/wd/hub"), chromeOptions.ToCapabilities());
            driver.Navigate().GoToUrl("http://www.google.com");
            driver.Quit();
```

Also with (/wd/hub and without /wd/hub)
```
var chromeOptions = new ChromeOptions();
           
            IWebDriver driver = new RemoteWebDriver(new Uri("http://127.0.0.1:4444/wd/hub"), chromeOptions.ToCapabilities());
            driver.Navigate().GoToUrl("http://www.google.com");
            driver.Quit();

```

**Comment by AceOfSpades29** [comment]
I can confirm that upgrading to .NET 5 I am able to create a session. Seems like this maybe only compatible with .NET Core 2.1 or higher. Not sure if it's mentioned in the doc anywhere if this is the case.

**Comment by jimevans** [comment]
It is not the intent to be "only compatible with .NET Core." The code being executed is identical in the bindings for .NET Framework and .NET Core, so if something isn't working in one and works in the other, it sounds like a framework difference that isn't documented or known about. I'm not sure what Selenium can do about that in its source code.

**Comment by jimevans** [comment]
Fixed in 46de1e6.
